### PR TITLE
add constructor to main widget 

### PR DESCRIPTION
**Pull request by vaniisgh**
Hey,
This issue ( #94 ) looked interesting and beginner-friendly,
I just wanted to know if this was a good start.

### CODE REVIEW

**Comment by ianhi** [comment]
Hi @vaniisgh thanks for taking this on! This looks like a good start but there are a few things to clean up.

TODO:

1. Add some examples documenting this
2. Add a docstring to `__init__` with  parameters section
3. Cleanup checking for string in the `from_json` method (I'll add an inline review)
4. Add a check for neo4j graphs
5. Change the name from `graph_file` to something more general.
    - It's not necessarily a file that user will be passing in.
6. Make the formatter happy by running `black .`



Also looking at the test failures there are unfortunately several due to neo4j which aren't your fault. This makes it trickier to see if your are causing issues but you can look more closely by clicking on the `details` next to the failing test.

File: ipycytoscape/cytoscape.py [hunk 1]
```diff
@@ -504,11 +507,15 @@ def add_graph_from_json(self, json_file, directed=False, multiple_edges=False):
 
         Parameters
         ----------
-        json_file : dict
+        json_file : dict, string
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:510]: Can you add a line under this explaining somethign to effect of:

`dict in defining the graph or path to a json file`

File: ipycytoscape/cytoscape.py [hunk 2]
```diff
@@ -504,11 +507,15 @@ def add_graph_from_json(self, json_file, directed=False, multiple_edges=False):
...
         directed : bool
             If True all edges will be given 'directed' as a class if
             they do not already have it.
         """
+        if path.isfile(str(json_file)):
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:515]: Does this fully protect us if the user passes in a `dict`?

File: ipycytoscape/cytoscape.py [hunk 3]
```diff
@@ -811,12 +818,21 @@ class CytoscapeWidget(DOMWidget):
 
     graph = Instance(Graph, args=tuple()).tag(sync=True, **widget_serialization)
 
-    def __init__(self, **kwargs):
+    def __init__(self, graph_file=None, **kwargs):
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:821]: I don't love the name `graph_file` as it implies that the user is working with a file. Can you:

1. Change the name to be `graph`
2. Also add a check if it is an `ipycytoscape.Graph` object and if so use it as it as the graph directly.

File: ipycytoscape/cytoscape.py [hunk 4]
```diff
@@ -3,12 +3,15 @@
...
 # Distributed under the terms of the Modified BSD License.
-
+from os import path
 import copy
+import json
 
 from spectate import mvc
 from traitlets import TraitType, TraitError
 
+from pandas import DataFrame
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:13]: We currently don't require `pandas` so this need to either be an optional import or the check below needs to check for `pandas` in the `graph.__class__`

File: ipycytoscape/cytoscape.py [hunk 5]
```diff
@@ -811,12 +818,21 @@ class CytoscapeWidget(DOMWidget):
...
 
+        if isinstance(graph_file, nx.Graph):
+            self.graph.add_graph_from_networkx(graph_file)
+        if isinstance(graph_file, dict):
+            self.graph.add_graph_from_json(graph_file)
+        if isinstance(graph_file, DataFrame):
+            self.graph.add_graph_from_df(graph_file)
+        if isinstance(graph_file, str):
+            self.graph.add_graph_from_json(graph_file)
+
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:835]: Combine the checks that lead to `json`

```suggestion
        if isinstance(graph_file, nx.Graph):
            self.graph.add_graph_from_networkx(graph_file)
        if isinstance(graph_file, (dict, str)):
            self.graph.add_graph_from_json(graph_file)
        if isinstance(graph_file, DataFrame):
            self.graph.add_graph_from_df(graph_file)
```

Comment (ianhi) [reply ipycytoscape/cytoscape.py:835]: Also need a check for `neo4j`

**Comment by vaniisgh** [comment]
hey thank you so much for the feedback!
I tried to take the things you mentioned into account and made a few changes though, I still think I might need to revise my work a bit. 
Also, I apologize forgot to rebase my work on the current master I hope I did it correctly now.

File: ipycytoscape/cytoscape.py (hunk 4 above)
Comment (vaniisgh) [reply ipycytoscape/cytoscape.py]: I made a change to fix this, but I am not sure I passed the kwargs correctly to the add_graph_from_df correctly ... cold you please let me know what the correct way to do that would be? :)

File: ipycytoscape/cytoscape.py [hunk 6]
```diff
@@ -814,12 +825,35 @@ class CytoscapeWidget(DOMWidget):
...
+        """
+        Initializes the graph widget.
+
+        Parameters
+        ----------
+        - graph:
+          type: string|dict|pandas.DataFrame|networkx.Graph|neo4j|Graph|None
+          description: graph passes is checked to be of one of the declared
+                       types, and the corresponding object is added to the
+                       graph attribute of the DOM object.
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:838]: Thanks for adding this! Can you update this to be in the numpydoc style as described here: https://numpydoc.readthedocs.io/en/latest/format.html#parameters

You can see an example here: https://github.com/cytoscape/ipycytoscape/blob/b80a91329e7214d6587a9840dbd73e4cb9882366/ipycytoscape/cytoscape.py#L450-L458

**Review by ianhi — COMMENTED** [review]
This is looking good! I'll try look more closely (e.g. at the pandas import) in the coming days. For now I left a quick comment on the docstring and approved the github actions.

File: ipycytoscape/cytoscape.py [hunk 7]
```diff
@@ -814,12 +825,35 @@ class CytoscapeWidget(DOMWidget):
...
 
         self.on_msg(self._handle_interaction)
         self.graph = Graph()
 
+        if isinstance(graph, nx.Graph):
+            self.graph.add_graph_from_networkx(graph)
+        elif isinstance(graph, (dict, str)):
+            self.graph.add_graph_from_json(graph)
+        elif graph.__class__.__name__ == "DataFrame":
+            self.graph.add_graph_from_df(graph, **kwargs)
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:850]: So this is the approach I've always taken to optional imports in the past, but looking at it now and reading around on the internet a bit I think this is probably not the best way. This is brittle to changes in class names and also to other objects that have a similar classname. I think a better approach would be to try to import `pandas` at the top level like this:


```python
try:
    import pandas as pd
except ImportError:
    pd = None
```

and then for this checks do:

```python
elif pd and isinstance(graph, pd.DataFrame)
```


what do you think?

File: ipycytoscape/cytoscape.py [hunk 8]
```diff
@@ -814,12 +825,35 @@ class CytoscapeWidget(DOMWidget):
...
+        else:
+            if isinstance(graph, Graph):
+                self.graph = graph
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:853]: any reason not to make these `elif`s at the same level?

Comment (vaniisgh) [reply ipycytoscape/cytoscape.py:853]: What I was thinking was, that the graph.__class__.__name__ is "Graph" for both objects, and I didn't want to add the optional import, I will recheck though.

File: ipycytoscape/cytoscape.py (hunk 7 above)
Comment (vaniisgh) [reply ipycytoscape/cytoscape.py:850]: this seems like a good solution really. :) Though then should we do something similar for the neo4j graphs?

Comment (ianhi) [reply ipycytoscape/cytoscape.py:850]: Yup that's the best way forward!

**Comment by vaniisgh** [comment]
I think it should be better now :)

**Comment by ianhi** [comment]
Thanks for the ping! This slipped through the cracks for me - can look later

File: ipycytoscape/cytoscape.py [hunk 9]
```diff
@@ -28,6 +31,19 @@
...
+    import pandas as pd
+
+except ImportError:
+    pd = None
+
+try:
+    import py2neo
+
+except ImportError:
+    pass
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:44]:
```suggestion
    py2neo = None
```

File: ipycytoscape/cytoscape.py [hunk 10]
```diff
@@ -814,12 +838,35 @@ class CytoscapeWidget(DOMWidget):
...
 
-    def __init__(self, **kwargs):
+    def __init__(self, graph=None, **kwargs):
+        """
+        Initializes the graph widget.
+
+        Parameters
+        ----------
+        graph: string or dict or pandas DataFrame object or networkx Graph
+               object or neo4j Graph object or Graph object, optional (defaults
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:848]:
```suggestion
        graph: string, dict, pandas.DataFrame, networkx.Graph, neo4j.Graph, Graph object, optional
```

File: ipycytoscape/cytoscape.py [hunk 11]
```diff
@@ -814,12 +838,35 @@ class CytoscapeWidget(DOMWidget):
...
+               to None)
+               Checked to be of one of the declared types, and graph object
+               corresponding to it is added as a Graph attribute to the
+               CytoscapeWidget DOM.
```
Comment (ianhi) [inline ipycytoscape/cytoscape.py:852]:
```suggestion
               The graph to initialize with. Equivalent to calling the appropriate ``CytoscapeWidget.graph.add_graph_from_` method.
```

Is this clearer? We maybe should stay away from using words like `DOM` as I suspect that isn't a super clear term to most python users.

**Comment by vaniisgh** [comment]
are there any other changes that I could make :) ?

**Comment by vaniisgh** [comment]
@ianhi should i close this ?

**Comment by ianhi** [comment]
Thank you @vaniisgh !! This is a great addition and we can always iterate more as needed.

**Comment by ianhi** [comment]
> @ianhi should i close this ?

Definitely not that - I'm really sorry about the long periods of non-responsiveness that's totally on me. I really appreciate you coming and putting the effort to make a positive change. I can't promise to be perfect, but I can promise that I will at least try to be more honest myself about what I can claim to do. Also don't be scared to ping more and more often (though ideally course you shouldn't have to :/ )

**Comment by ianhi** [comment]
This should be live in version `1.3.0` soon!

**Pull request merged**